test/e2e/helpers/channel_defaulter_test_helper.go

```diff
@@ -0,0 +1,180 @@
+		client := common.Setup(st, false)
+		defer common.TearDown(client)
+
+		if err := updateDefaultChannelCM(client, func(conf *defaultchannel.Config) {
+			setNamespaceDefaultChannel(conf, client.Namespace, channel)
+		}); err != nil {
+			st.Fatalf("Failed to update the defaultchannel configmap: %v", err)
+		}
+
+		defaultChannelTestHelper(st, client, channel)
+	})
+}
+
+func defaultChannelTestHelper(t *testing.T, client *common.Client, expectedChannelKind string) {
+	channelName := "e2e-defaulter-channel"
+	senderName := "e2e-defaulter-sender"
+	subscriptionName := "e2e-defaulter-subscription"
+	loggerPodName := "e2e-defaulter-logger-pod"
+
+	// create channel
+	client.CreateChannelWithDefaultOrFail(eventingtesting.NewChannel(channelName, client.Namespace))
+
+	// create logger service as the subscriber
+	pod := resources.EventLoggerPod(loggerPodName)
+	client.CreatePodOrFail(pod, common.WithService(loggerPodName))
+
+	// create subscription to subscribe the channel, and forward the received events to the logger service
+	client.CreateSubscriptionOrFail(
+		subscriptionName,
+		channelName,
+		common.ChannelTypeMeta,
+		resources.WithSubscriberForSubscription(loggerPodName),
+	)
+
+	// wait for all test resources to be ready, so that we can start sending events
+	if err := client.WaitForAllTestResourcesReady(); err != nil {
+		t.Fatalf("Failed to get all test resources ready: %v", err)
+	}
+
+	// check if the defaultchannel creates exactly one underlying channel given the spec
+	metaResourceList := resources.NewMetaResourceList(client.Namespace, common.GetChannelTypeMeta(expectedChannelKind))
+	objs, err := base.GetGenericObjectList(client.Dynamic, metaResourceList, &eventingduck.SubscribableType{})
+	if err != nil {
+		t.Fatalf("Failed to list the underlying channels: %v", err)
+	}
+	if len(objs) != 1 {
+		t.Fatalf("The defaultchannel is expected to create 1 underlying channel, but got %d", len(objs))
+	}
+
+	// send fake CloudEvent to the channel
+	body := fmt.Sprintf("TestSingleEvent %s", uuid.NewUUID())
+	event := &resources.CloudEvent{
+		Source:   senderName,
+		Type:     resources.CloudEventDefaultType,
+		Data:     fmt.Sprintf(`{"msg":%q}`, body),
+		Encoding: resources.CloudEventDefaultEncoding,
+	}
+
+	if err := client.SendFakeEventToAddressable(senderName, channelName, common.ChannelTypeMeta, event); err != nil {
+		t.Fatalf("Failed to send fake CloudEvent to the channel %q", channelName)
+	}
+
+	// verify the logger service receives the event
+	if err := client.CheckLog(loggerPodName, common.CheckerContains(body)); err != nil {
+		t.Fatalf("String %q not found in logs of logger pod %q: %v", body, loggerPodName, err)
+	}
+}
+
+// updateDefaultChannelCM will update the default channel configmap
+func updateDefaultChannelCM(client *common.Client, updateConfig func(config *defaultchannel.Config)) error {
+	systemNamespace := resources.SystemNamespace
```

Comment:
Do you need to do all these programmatic reading and updating of the configMap? and then call the defaulter UpdateConfigMap? Can't you just create a ConfigMap object in memory and set the appropriate value to a key in the Data field?
No need to do it now, but maybe when you add the thread-safety thing...

Reply:
Yeah I can do that, but I think we still cannot get rid of the `yaml.Unmarshal` and `yaml.Marshal` stuff, since it's not quite safe to directly change yaml, even in test.